hal/tests/TESTS/pin_names/pinvalidate.py

```diff
@@ -0,0 +1,453 @@
+
+    return tabulate(table, headers=['Platform name', 'Test suite', 'Test case', 'Result', 'Error count', 'Errors'], tablefmt=tablefmt)
+
+def print_pretty_html_report(report):
+    output  = "<html><head><style>table, td, tr { border: 2px solid black; border-collapse: collapse; padding: 5px; font-family: Helvetica, serif; }</style></head>"
+    output += "<body><p><button onclick='e=document.getElementsByTagName(\"details\");for(var i=0;i<e.length;i++){e[i].setAttribute(\"open\",\"true\")};'>Expand all errors</button></p><table>"
+    output += "<tr><th>Platform name</th><th>Test suite</th><th>Test case</th><th>Result</th><th>Error count</th><th>Errors</th></tr>"
+    
+    for case in report:
+        output += "<tr>"
+
+        if case['errors']:
+            error_details = "<details><summary>View errors</summary><table>"
+            for error in case['errors']:
+                error_details += "<tr>"
+                if error['key']:
+                    error_details += "<td>" + error['key'] + "</td>"
+                if error['val']:
+                    error_details += "<td>" + error['val'] + "</td>"
+                if error['message']:
+                    error_details += "<td>" + error['message'] + "</td>"
+                error_details += "</tr>"
+            error_details += "</table></details>"
+        else:
+            error_details = ""
+        
+        output += "<td>" + case['platform_name'] + "</td>"
+        output += "<td>" + case['suite_name'] + "</td>"
+        output += "<td>" + case['case_name'] + "</td>"
+
+        if case['result'] == "PASSED":
+            color = "green"
+            count_color = "black"
+        else:
+            color = "red"
+            count_color = "red"
+        
+        output += "<td style='color:" + color + "'>" + case['result'] + "</td>"
+        output += "<td style='color:" + count_color + "'>" + str(len(case['errors'])) + "</td>"
+        output += "<td>" + error_details + "</td>"
+
+        output += "</tr>"
+
+    output += "</table></body></table>"
+
+    return output
+
+def all_cases_passed(report):
+    result = True
+    for case in report:
+        if case['result'] == 'FAILED':
+            result = False
+    return result 
+
+test_cases = [
+    {'suite_name': 'generic', 'case_name': 'identity', 'case_function': identity_assignment_check, 'case_input': 'dict'},
+    {'suite_name': 'generic', 'case_name': 'nc', 'case_function': nc_assignment_check, 'case_input': 'dict'},
+    {'suite_name': 'generic', 'case_name': 'duplicate', 'case_function': duplicate_assignment_check, 'case_input': 'dict'},
+    {'suite_name': 'generic', 'case_name': 'legacy', 'case_function': legacy_assignment_check, 'case_input': 'content'},
+    {'suite_name': 'arduino', 'case_name': 'duplicate', 'case_function': arduino_duplicate_assignment_check, 'case_input': 'dict'},
```

Comment:
It might be worth being explicit about "arduino_uno" tests as opposite to the legacy "arduino" solution

Reply:
This might be more a cosmetic change, which may not care. @gpsimenos up to you.

Reply:
I think we have to change for arduino_uno

Reply:
> > For arduino support, I'm not sure if there's a simple way to check in Travis. We'd probably run into the same issues that `pinvalidate.py` has with matching `PinNames.h` files to targets in `targets.json`.
> 
> Means that arduino_uno pins will not be checked by travis ?

That's up to us to decide whether we will manually enable the arduino_uno test suite or not. But if we do enable it, targets that don't have arduino form factor will show failures.

An alternative idea is to add the following logic to the script: if at least one ARDUINO_UNO_XX value is defined, consider the target has arduino uno form factor and run the arduino_uno test suite. If no ARDUINO_UNO_XX values are defined at all, don't run the arduino_test suite.

How does that sound?

Reply:
Done